### PR TITLE
launcher

### PR DESCRIPTION
**Pull request by bartoszkobylinski**

### CODE REVIEW

**Comment by busterbeam** [comment]
You'll need to fix some things before closing the branch. as the linting checks aren't being accepted.  `pre-commit` should of caught this. but also getting pyright again. you can ignore pyright for now.

flake8 is totally right about docstrings.  but besides that this code looks superb I'll test run it in an hour.

File: jigsaw-game/client/menu.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    view_functions = {
+        "menu": menu_view,
+        "game_selection": game_selection_view,
+    }
+
+    while True:
+        if current_view == "play_game":
+            action = play_game(current_difficulty)
+        else:
+            action = view_functions[current_view]()
```
Comment (busterbeam) [inline jigsaw-game/client/menu.py:94]: interesting usage of accessing functions.  maybe instead try `globals()[current_view]()`.. but maybe add a few lines, to check we are accessing a function such as `function = globals()[current_view] if callable(globals()[current_view]) else not_implemented_function`.

File: jigsaw-game/client/menu.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    return False
+
+
+# Views
+def game_selection_view():
+    screen.fill(WHITE)
+
+    easy_button = pygame.Rect(100, 100, 200, 50)
+    moderate_button = pygame.Rect(100, 200, 200, 50)
+    hard_button = pygame.Rect(100, 300, 200, 50)
```
Comment (busterbeam) [inline jigsaw-game/client/menu.py:38]: for **Quick Game** Easy, Intermediate and Hard are good options, but for **Custom Game** we should include either text entry or better still a slider for how many "tiles"... but make it follow squared numbers 4, 9, 16, 25, 36, 49, 64, 81, etc.

File: jigsaw-game/client/menu.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+WHITE = (255, 255, 255)
+RED = (255, 0, 0)
+BLUE = (0, 0, 255)
+
+# Screen dimensions
+WIDTH, HEIGHT = 800, 600
+screen = pygame.display.set_mode((WIDTH, HEIGHT))
+pygame.display.set_caption('Game Menu')
+
+font = pygame.font.SysFont(None, 36)  # Default font, size 36
```
Comment (busterbeam) [inline jigsaw-game/client/menu.py:16]: this should be in `menu` function and dimensions could be put into a `constants.py` file as `GameClient` uses it aswell, but we should resizeable, and this should probably a "configurable" option in "Settings".. do have fun with fonts

**Review by busterbeam — CHANGES_REQUESTED** [review]
good start, hopefully this will help

File: jigsaw-game/client/menu.py (hunk 1 above)
Comment (LukeHankey) [reply jigsaw-game/client/menu.py:94]: I would argue strongly against using `globals()`. It's a bad practice to get into a habit of. Simple if checks on all of them is fine, or to clean it up in one line, just have `action = view_functions[current_view]()` and add the `play_game` action to `view_functions`.

Comment (busterbeam) [reply jigsaw-game/client/menu.py:94]: :smirk: don't like the dangerous side? as long as we are checking that they are callable. also a simple startswith filter for dunders would probably fix the more extreme cases.  Either that, or really it should be using `function.__name__` in the key place so you don't have to update code everywhere when you change names.

File: jigsaw-game/client/buttons.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
+import pygame
+import pygame_gui
```
Comment (LukeHankey) [inline jigsaw-game/client/buttons.py:2]: Add to `requirements.txt`

File: jigsaw-game/client/buttons.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+from constants import HEIGHT, WIDTH
```
Comment (LukeHankey) [inline jigsaw-game/client/buttons.py:3]:
```suggestion
from .constants import HEIGHT, WIDTH
```

File: jigsaw-game/client/menu.py [hunk 6]
```diff
@@ -167,7 +220,6 @@ def menu():
         pygame.display.flip()
 
     pygame.quit()
-    return action
 
-if __name__ == "__main__":
-    print(menu())
+
+menu()
```
Comment (LukeHankey) [inline jigsaw-game/client/menu.py:225]:
```suggestion
```

File: jigsaw-game/client/settings_view.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
+import pygame.event
+import pygame_gui.elements
+from pygame import QUIT
+
+from constants import screen, WHITE, RED, BLUE, font, WIDTH, HEIGHT, manager
+
+#  Create a GUI components
+'''
```
Comment (LukeHankey) [inline jigsaw-game/client/settings_view.py:8]: What is this commented for?

File: jigsaw-game/client/settings_view.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
+import pygame.event
+import pygame_gui.elements
+from pygame import QUIT
+
+from constants import screen, WHITE, RED, BLUE, font, WIDTH, HEIGHT, manager
```
Comment (LukeHankey) [inline jigsaw-game/client/settings_view.py:5]:
```suggestion
from .constants import screen, WHITE, BLUE, font, WIDTH, manager
```

**Review by LukeHankey — CHANGES_REQUESTED** [review]

**Comment by PiLogic** [comment]
Hello! Is there any chance my username could be added to the Credits Page? Thank you, that would be greatly appreciated.

**Pull request merged**